Support RollUpApply and PathExpression in slotted runtime

henriknyman opened the pull request.

SaschaPeukert committed:
Started implementing RollUpApplySlottedPipe

henriknyman committed:
Support path expression in slotted runtime

- Make PathStep rewritable so that it's variables can be reached by
slotted rewriter
- Add slotted version of ProjectedPath
- Slot allocation of RollUpApply
- Remove passing tests from blacklists

henriknyman committed:
Precompute more in RollUpApplySlottedPipe

- Minor cleanup

henriknyman committed:
Move slotted compatibility folder from v3_3 to v3_4

henriknyman committed:
Update tests that are now running with slotted runtime

henriknyman committed:
Fix license headers

henriknyman committed:
Fix slot allocation of RollUpApply

- Allocate slots based on lhs context and discard rhs
after collecting results
- Fix issue with initial context on rhs
- Support collection of Node and Relationship values

systay commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/pipes/RollUpApplySlottedPipe.scala:
```diff
@@ -0,0 +1,84 @@
+    (state: QueryState) => (ctx: ExecutionContext) => expression(ctx, state)
+  }
+
+  private val hasNullValuePredicates: Seq[(ExecutionContext) => Boolean] =
```
I think I would prefer this to be done outside this file and have this file focus on the runtime aspects. When reading this code I want to be in a runtime frame of mind, and here we are mixing between the two.

Pretty sure this is highly subjective, so feel free to ignore me. This is just my humble opinion.

henriknyman replied:
I don't really have a strong opinion, but I chose to humbly ignore yours :) Currently there is a mix of two styles across slotted pipe implementations, so maybe we could agree on one style, document it and unify the code as a separate concern.

systay commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/pipes/RollUpApplySlottedPipe.scala:
```diff
@@ -0,0 +1,84 @@
+    nullableIdentifiers.toSeq.map { elem =>
+      val elemSlot = pipelineInformation.get(elem)
+      elemSlot match {
+        case Some(LongSlot(offset, true, _, _)) => { (ctx: ExecutionContext) => ctx.getLongAt(offset) == -1 }
```
We know which of these slots are nullable. No need to check for null if it is not nullable.

systay commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/pipes/RollUpApplySlottedPipe.scala:
```diff
@@ -0,0 +1,84 @@
+  override protected def internalCreateResults(input: Iterator[ExecutionContext], state: QueryState) = {
+    input.map {
+      ctx =>
+        val outputRow = PrimitiveExecutionContext(pipelineInformation)
```
No need to create a new row and copy the incoming data into it. We should just write to the incoming `ctx`.

systay commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/pipes/RollUpApplySlottedPipe.scala:
```diff
@@ -0,0 +1,84 @@
+          val innerState = state.withInitialContext(ctx)
+          val innerResults: Iterator[ExecutionContext] = rhs.createResults(innerState)
+          val collection = VirtualValues.list(innerResults.map(getValueToCollectFunction(state)).toArray: _*)
+          setCollectionInRow(outputRow, collection)
```
Instead of using a method value, I think we should calculate the offset to an int at constructor time and just `outputRow.setRefAt` here. This is in the tight loop, so minimizing virtual calls is a good idea.

systay commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/SlottedPipeBuilder.scala:
```diff
@@ -274,28 +274,28 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
     runtimeColumns
   }
 
+  private def createProjectionForIdentifier(pipelineInformation1: PipelineInformation)(identifier: String) = {
```
How about just `slotInformation`? That `1` itches a little... ;)

henriknyman replied:
It was already there, I just extracted a method :) Renaming it something different just because there is an outer instance with the same type seems equally weird IMHO.

systay commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocation.scala:
```diff
@@ -271,12 +271,17 @@ object SlotAllocation {
-        cartesianProductPipeline
+        newPipeline
+
+      case RollUpApply(_, _, collectionName, _, _) =>
```
Could we beef up `SlotAllocationTest` to match? We've been pretty good with the test coverage there, would be nice to keep it up.

henriknyman replied:
Yes

systay commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocation.scala:
```diff
@@ -271,12 +271,17 @@ object SlotAllocation {
+        newPipeline
+
+      case RollUpApply(_, _, collectionName, _, _) =>
+        val newPipeline = lhsPipeline.seedClone()
```
I don't think we need to do a `seedClone()` here. Pretty sure we should just add one reference slot to the `lhs` pipeline. For each incoming row, we're producing exactly one row, and the only difference between the incoming and outgoing row is that the outgoing one has one additional refslot, nothing else.

fickludd replied:
yes, l think `lhs` should continue as well.

systay commented on enterprise/cypher/slotted-runtime/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/pipes/RollUpApplyPipeTest.scala:
```diff
@@ -0,0 +1,185 @@
+import org.neo4j.values.storable.Values.NO_VALUE
+import org.neo4j.values.virtual.VirtualValues
+
+class RollUpApplySlottedPipeTest extends CypherFunSuite with PipeTestSupport with FakeEntityTestSupport {
```
Nice test coverage! 👍

systay reviewed (changes requested):
Nice work. I think we are there functionality-wise. My comments are mostly performance improvements.

fickludd commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/expressions/SlottedExpressionConverters.scala:
```diff
@@ -61,7 +63,44 @@ object SlottedExpressionConverters extends ExpressionConverter {
+    }
+
+    val projector = project(e.step)
+//    val dependencies = e.step.dependencies.map(_.asInstanceOf[Variable].name) // TODO: Verify that we never need this after slot rewriting
```
There a TODO here. Is that acted on, or cardified?

fickludd reviewed (changes requested):
From what I understand, this looks good in general. As andres point out, we could avoid copying rows from the left though, which would be nice to fix.

henriknyman committed:
Fix compilation errors after rebase

henriknyman committed:
Fix some review comments

- Use lhs pipeline
- Assume collection is always a RefSlot and use int offset
- Add a test to SlotAllocationTest

henriknyman committed:
Fix symbol table dependencies in SlottedProjectedPath

- Update succeeding acceptance tests
